### PR TITLE
8051725: Improve expansion of Conv2B nodes in the middle-end

### PR DESCRIPTION
**Pull request by jaskarth**
Hi, I've created optimizations for the expansion of `Conv2B` nodes, especially when followed immediately by an xor of 1. This pattern is fairly common, and can arise from both [cmov idealization](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/movenode.cpp#L241) and [diamond-phi optimization](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/cfgnode.cpp#L1571). This change replaces `Conv2B` nodes in the middle-end during post loop opts IGVN with conditional moves on supported platforms (x86_64, aarch64, arm32), allowing the bit flip with `xor` to be subsumed with an inversion of the comparison instead. This change also reduces the overhead of the matcher in the backends, as fewer rules need to be traversed in order to match an ideal node. Performance results from my (Zen 2) machine:

```
                                    Baseline                           Patch              Improvement
Benchmark                      Mode  Cnt  Score    Error Units     Score    Error  Units
Conv2BRules.testEquals0        avgt   10  47.566 ± 0.346 ns/op  /  34.130 ± 0.177  ns/op  + 28.2%
Conv2BRules.testNotEquals0     avgt   10  37.167 ± 0.211 ns/op  /  34.185 ± 0.258  ns/op  + 8.0%
Conv2BRules.testEquals1        avgt   10  35.059 ± 0.280 ns/op  /  34.847 ± 0.160  ns/op  (unchanged)
Conv2BRules.testEqualsNull     avgt   10  56.768 ± 2.600 ns/op  /  34.330 ± 0.625  ns/op  + 39.5%
Conv2BRules.testNotEqualsNull  avgt   10  47.447 ± 1.193 ns/op  /  34.142 ± 0.303  ns/op  + 28.0%
```
Reviews would be greatly appreciated!

Testing: tier1-2 on linux x64, GHA

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8051725](https://bugs.openjdk.org/browse/JDK-8051725): Improve expansion of Conv2B nodes in the middle-end

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Quan Anh Mai](https://openjdk.org/census#qamai) (@merykitty - Committer) ⚠️ Review applies to [69e914a5](https://git.openjdk.org/jdk/pull/13345/files/69e914a5efb8d46209aeca5d326706abe2536831)
 * [Sandhya Viswanathan](https://openjdk.org/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13345/head:pull/13345` \
`$ git checkout pull/13345`

Update a local copy of the PR: \
`$ git checkout pull/13345` \
`$ git pull https://git.openjdk.org/jdk.git pull/13345/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13345`

View PR using the GUI difftool: \
`$ git pr show -t 13345`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13345.diff">https://git.openjdk.org/jdk/pull/13345.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13345#issuecomment-1496924387)

### CODE REVIEW

**Comment by merykitty** [comment]
I think this should be done in the middle-end instead. May I ask what are the advantages of `Conv2B` over `CMove` that we need to have it all the way to matching?

**Comment by jaskarth** [comment]
As far as I know, `Conv2B` is a special-case convert node where it does either `c == 0 ? 0 : 1` or `p == null ? 0 : 1`. I think an advantage of Conv2B over CMove here is that the Conv2B has more specialized rules for value() and identity(), so it can prune more types of inputs than an equivalent Cmove can. I agree that it's not great having to shuffle the node from the middle-end to the backend, but I think it's still helpful as it can remove some dead bools that CMove wouldn't be able to. Hope this clarifies a bit!

**Comment by merykitty** [comment]
@jaskarth Yes I also think that is the case, but without advantages in the back-end maybe it's best us lowering it in macro expansion phase so the compiler can have chances to transform more primitive `CMove`, what do you think?

**Comment by jaskarth** [comment]
Oh, I had forgotten to consider the macro expansion step! That sounds reasonable to me, I'll make this change and see what the performance is like.

**Comment by jaskarth** [comment]
I've reworked the transformation to happen in macro expansion, and it seems the performance is actually *better* now!
```
                                    Baseline                           Patch              Improvement
Benchmark                      Mode  Cnt  Score    Error Units     Score    Error  Units
Conv2BRules.testEquals0        avgt   10  47.566 ± 0.346 ns/op  /  34.130 ± 0.177  ns/op  + 28.2%
Conv2BRules.testNotEquals0     avgt   10  37.167 ± 0.211 ns/op  /  34.185 ± 0.258  ns/op  + 8.0%
Conv2BRules.testEquals1        avgt   10  35.059 ± 0.280 ns/op  /  34.847 ± 0.160  ns/op  (unchanged)
Conv2BRules.testEqualsNull     avgt   10  56.768 ± 2.600 ns/op  /  34.330 ± 0.625  ns/op  + 39.5%
Conv2BRules.testNotEqualsNull  avgt   10  47.447 ± 1.193 ns/op  /  34.142 ± 0.303  ns/op  + 28.0%
```
The comparison is now basically the same as doing it with `cmp`, which is nice. It seems the reason is because the assembly now zeroes the register, tests against zero, and then does the `setcc`, instead of comparsion, `setcc`, then `movzbl`. So, it seems that doing the transform in macro expansion is indeed a better choice for x86, as well as reducing the overhead in the matcher. However, I'm not so sure if the benefit will be the same across other platforms as it seems like the different architectures implement `Conv2B` using different strategies. Do you have any thoughts on this approach @merykitty?

**Review by TobiHartmann — COMMENTED** [review]
I'm a bit confused, why do you need the new match rules if Conv2B nodes are now macro expanded to CMove?

**Comment by jaskarth** [comment]
Hi, that's my bad- I was working on removing the now-redundant Conv2B rules from the backends but I hadn't had a chance to push that yet, I've done so now. Since this has become a broader cleanup effort I'll also go ahead and rename the JBS issue to reflect the new approach with macro expansion as well.

**Comment by merykitty** [comment]
Generally I think it's good, one small question I want to ask is whether we should do splitting `Xor` through `CMove` as an idealisation of the `Xor`?

**Comment by jaskarth** [comment]
I think that's a good idea, it would reduce the complexity of the logic in macro expansion and allow the transform to be applied more generally.

**Review by RealFYang — COMMENTED** [review]
Hello, I wonder if we could make this transformation of Conv2B conditional? Architectures like RISC-V doesn't have support of conditional moves at the ISA level for now. So we set ConditionalMoveLimit parameter to 0 for this platform and conditionals moves are emulated with normal compare and branch instructions instead [1]. I don't think we would achieve better performance numbers on this platform with this change.

[1] https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/riscv/riscv.ad#L9600

**Comment by jaskarth** [comment]
Hey @RealFYang, thanks for this info! I wasn't aware that RISC-V didn't have conditional moves, and I agree that it doesn't sound like this transform would be so profitable there. To make the transformation conditional I've moved it to post loop opts IGVN, and only run it if the match rule for `Conv2B` isn't found. In an effort to not accidentally cause performance regressions, I've limited the transform to x86_64, aarch64, and arm32.

@merykitty I've also implemented your change with idealization and would like your thoughts on it, thanks!

I'll attach aarch64 perf results soon.

**Comment by RealFYang** [comment]
@jaskarth : Thanks for taking care of that. I performed tier1-3 tests on linux-riscv64 platform, result looks good.

File: src/hotspot/share/opto/macro.cpp [hunk 1]
```diff
@@ -41,6 +41,7 @@
 #include "opto/loopnode.hpp"
 #include "opto/macro.hpp"
 #include "opto/memnode.hpp"
+#include "opto/movenode.hpp"
```
Comment (merykitty) [inline src/hotspot/share/opto/macro.cpp:44]: Unnecessary change?

File: src/hotspot/share/opto/cfgnode.cpp [hunk 2]
```diff
@@ -1567,7 +1573,7 @@ static Node *is_x2logic( PhaseGVN *phase, PhiNode *phi, int true_path ) {
   // Build int->bool conversion
   Node *n = new Conv2BNode(cmp->in(1));
   if( flipped )
-    n = new XorINode( phase->transform(n), phase->intcon(1) );
+  n = new XorINode( phase->transform(n), phase->intcon(1) );
```
Comment (merykitty) [inline src/hotspot/share/opto/cfgnode.cpp:1576]: This lives under the `if (flipped)`, maybe move into a block for more clarity.

File: src/hotspot/share/opto/addnode.cpp [hunk 3]
```diff
@@ -886,6 +886,19 @@ Node* XorINode::Ideal(PhaseGVN* phase, bool can_reshape) {
       phase->record_for_igvn(this);
     }
   }
+
+  // Try to convert (c ? 1 : 0) ^ 1 into !c ? 1 : 0. This pattern can occur after expansion of Conv2B nodes.
```
Comment (merykitty) [inline src/hotspot/share/opto/addnode.cpp:890]: Be more general? `Xor (CMove cond, iftrue, iffalse), op == CMove cond, (Xor iftrue op), (Xor iffalse op)`. You can be conservative and apply this only if `op`, `iftrue` and `iffalse` are all constant.

File: src/hotspot/share/opto/macro.cpp (hunk 1 above)
Comment (jaskarth) [reply src/hotspot/share/opto/macro.cpp:44]: Oops, I forgot to remove the include from the old macro expansion logic. Thanks for the catch!

File: src/hotspot/share/opto/cfgnode.cpp (hunk 2 above)
Comment (jaskarth) [reply src/hotspot/share/opto/cfgnode.cpp:1576]: Fixed, thanks.

File: src/hotspot/share/opto/addnode.cpp (hunk 3 above)
Comment (jaskarth) [reply src/hotspot/share/opto/addnode.cpp:890]: I think that's a good idea, I've made this change. I wonder if other associative operations would also benefit from a similar patch?

File: src/hotspot/share/opto/cfgnode.cpp [hunk 4]
```diff
@@ -1523,6 +1523,12 @@ Node* PhiNode::unique_input(PhaseTransform* phase, bool uncast) {
 // Convert Phi to an ConvIB.
 static Node *is_x2logic( PhaseGVN *phase, PhiNode *phi, int true_path ) {
   assert(true_path !=0, "only diamond shape graph expected");
+
+  // If we're late in the optimization process, we may have already expanded Conv2B nodes
+  if (phase->C->post_loop_opts_phase()) {
+    return nullptr;
+  }
```
[the remark this answers is not in the document]
Comment (jaskarth) [reply src/hotspot/share/opto/cfgnode.cpp:1530]: Yes, I think adding that would help in not accidentally removing optimization opportunities here, for other platforms. Thanks for this!

File: src/hotspot/share/opto/convertnode.hpp [hunk 5]
```diff
@@ -33,10 +33,11 @@
 // Convert int/pointer to a Boolean.  Map zero to zero, all else to 1.
 class Conv2BNode : public Node {
   public:
-  Conv2BNode( Node *i ) : Node(0,i) {}
+  Conv2BNode(Node* i) : Node(nullptr, i) {}
```
[the remark this answers is not in the document]
Comment (jaskarth) [reply src/hotspot/share/opto/convertnode.hpp:36]: Thanks for the catch!

**Comment by jaskarth** [comment]
Apologies for the delayed update, and thanks for the reviews!

I have aarch64 performance results, from an M1 mac:
```
                                    Baseline                           Patch              Improvement
Benchmark                      Mode  Cnt  Score    Error Units     Score    Error  Units
Conv2BRules.testEquals0        avgt   12  41.697 ± 0.127 ns/op  /  40.724 ± 0.086  ns/op  + 2.4%
Conv2BRules.testNotEquals0     avgt   12  39.522 ± 0.143 ns/op  /  40.608 ± 0.046  ns/op  - 2.7%
Conv2BRules.testEquals1        avgt   12  40.168 ± 0.136 ns/op  /  40.679 ± 0.044  ns/op  (unchanged)
Conv2BRules.testEqualsNull     avgt   12  48.922 ± 0.498 ns/op  /  42.046 ± 0.018  ns/op  + 15.1%
Conv2BRules.testNotEqualsNull  avgt   12  41.725 ± 0.264 ns/op  /  42.063 ± 0.043  ns/op  - 0.8%
```
It seems like the patch doesn't have much of an impact other than `testEqualsNull`, which would make sense as the Conv2B rule is using the same `cset` instruction as the 0 and 1 rule for CMoveI. I was unfortunately not able to test for arm32, but I think it should still be beneficial as the Conv2B rules there used two cmoves and had a fixme, whereas with this patch it would only use one cmove.

**Comment by merykitty** [comment]
A tiny point: maybe just remove `setne` and use `setb(Assembler::notZero`, I don't think having a dedicated `setne` achieve much.

File: src/hotspot/share/opto/addnode.cpp [hunk 6]
```diff
@@ -886,6 +886,23 @@ Node* XorINode::Ideal(PhaseGVN* phase, bool can_reshape) {
...
+  if (in1->Opcode() == Op_CMoveI && in2->is_Con()) {
+    if (in1->in(2)->is_Con() && in1->in(3)->is_Con()) {
+      int cmp_op = in1->in(1)->in(1)->Opcode();
+
+      int in2_val = phase->type(in2)->is_int()->get_con();
+      int l_val = phase->type(in1->in(2))->is_int()->get_con();
+      int r_val = phase->type(in1->in(3))->is_int()->get_con();
+
+      if (cmp_op == Op_CmpI || cmp_op == Op_CmpP) {
+        // Flip the sense of comparison in the bool and return a new cmove
```
[the remark this answers is not in the document]
Comment (jaskarth) [reply src/hotspot/share/opto/addnode.cpp]: Fixed

**Comment by jaskarth** [comment]
Thanks for your review! I've updated the code accordingly. I also noticed that in addition to `setne`, there was also `setl` and `sete` with just a handful of uses, so I've modified those to also use `setb` with a condition code instead as well.

**Comment by TobiHartmann** [comment]
I run this through some quick testing and I'm seeing the following crash with `compiler/vectorapi/TestMaskedMacroLogicVector.java` and `-XX:-TieredCompilation -XX:+AlwaysIncrementalInline`:

```
#  Internal Error (/workspace/open/src/hotspot/share/opto/type.hpp:1967), pid=3563106, tid=3563120
#  assert(_base == Int) failed: Not an Int
#
# JRE version: Java(TM) SE Runtime Environment (21.0) (fastdebug build 21-internal-LTS-2023-05-23-0846313.tobias.hartmann.jdk2)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (fastdebug 21-internal-LTS-2023-05-23-0846313.tobias.hartmann.jdk2, mixed mode, sharing, compressed oops, compressed class ptrs, g1 gc, linux-amd64)
# Problematic frame:
# V  [libjvm.so+0x5cca5f]  Type::is_int() const [clone .part.0]+0x2f
#

Current CompileTask:
C2:   9002  777 %  b        compiler.vectorapi.TestMaskedMacroLogicVector::verifyInt2 @ 3 (136 bytes)

Stack: [0x00007f83d89e0000,0x00007f83d8ae1000],  sp=0x00007f83d8adb530,  free space=1005k
Native frames: (J=compiled Java code, j=interpreted, Vv=VM code, C=native code)
V  [libjvm.so+0x5cca5f]  Type::is_int() const [clone .part.0]+0x2f  (type.hpp:1967)
V  [libjvm.so+0x5d2488]  XorINode::Ideal(PhaseGVN*, bool)+0x468  (node.hpp:394)
V  [libjvm.so+0x151c7fe]  PhaseIterGVN::transform_old(Node*)+0x22e  (phaseX.cpp:833)
V  [libjvm.so+0x1516141]  PhaseIterGVN::optimize()+0x81  (phaseX.cpp:1218)
V  [libjvm.so+0x9f3a12]  PhaseIdealLoop::optimize(PhaseIterGVN&, LoopOptsMode)+0x4d2  (loopnode.hpp:1203)
V  [libjvm.so+0x9efdae]  Compile::Optimize()+0x10fe  (compile.cpp:2350)
V  [libjvm.so+0x9f24c5]  Compile::Compile(ciEnv*, ciMethod*, int, Options, DirectiveSet*)+0x1ae5  (compile.cpp:839)
V  [libjvm.so+0x84c414]  C2Compiler::compile_method(ciEnv*, ciMethod*, int, bool, DirectiveSet*)+0x3c4  (c2compiler.cpp:118)
V  [libjvm.so+0x9fe2c0]  CompileBroker::invoke_compiler_on_method(CompileTask*)+0xa00  (compileBroker.cpp:2265)
V  [libjvm.so+0x9ff148]  CompileBroker::compiler_thread_loop()+0x618  (compileBroker.cpp:1944)
V  [libjvm.so+0xeb96fc]  JavaThread::thread_main_inner()+0xcc  (javaThread.cpp:719)
V  [libjvm.so+0x179b59a]  Thread::call_run()+0xba  (thread.cpp:217)
V  [libjvm.so+0x14980cc]  thread_native_entry(Thread*)+0x11c  (os_linux.cpp:775)
```

File: src/hotspot/share/opto/addnode.cpp [hunk 7]
```diff
@@ -886,6 +886,22 @@ Node* XorINode::Ideal(PhaseGVN* phase, bool can_reshape) {
...
+      int cmp_op = in1->in(CMoveNode::Condition)->in(1)->Opcode();
+
+      int in2_val = phase->type(in2)->is_int()->get_con();
+      int l_val = phase->type(in1->in(CMoveNode::IfFalse))->is_int()->get_con();
+      int r_val = phase->type(in1->in(CMoveNode::IfTrue))->is_int()->get_con();
+
+      if (cmp_op == Op_CmpI || cmp_op == Op_CmpP) {
+        return new CMoveINode(in1->in(CMoveNode::Condition), phase->intcon(l_val ^ in2_val), phase->intcon(r_val ^ in2_val), TypeInt::INT);
+      }
+    }
```
Comment (sviswa7) [inline src/hotspot/share/opto/addnode.cpp:902]: An isa_int() check is needed before doing is_int()->get_con(). Something like below:
    ```
  const TypeInt* in2type = phase->type(in2)->isa_int();
      const TypeInt* ltype = phase->type(in1->in(CMoveNode::IfFalse))->isa_int();
      const TypeInt* rtype = phase->type(in1->in(CMoveNode::IfTrue))->isa_int();

      if (in2type && ltype && rtype) {
        int in2_val = in2type->get_con();
        int l_val = ltype->get_con();
        int r_val = rtype->get_con();

        if (cmp_op == Op_CmpI || cmp_op == Op_CmpP) {
          return new CMoveINode(in1->in(CMoveNode::Condition),
                                phase->intcon(l_val ^ in2_val),
                                phase->intcon(r_val ^ in2_val), TypeInt::INT);
        }
      }
```
This should fix the crash seen in vectorapi test.

Comment (jaskarth) [reply src/hotspot/share/opto/addnode.cpp]: Ah yep, I missed that when writing the function, and it's a simple fix. Thanks for helping diagnose the issue!

**Comment by jaskarth** [comment]
I think the latest change should fix the error! The code now properly checks that both types are ints, and tier1 and `hotspot_vector_1` testing passes without errors for me.

**Review by TobiHartmann — APPROVED** [review]
Great work, the latest version looks good to me. I'll run some testing and report back once it finished.

**Comment by TobiHartmann** [comment]
All tests passed.

**Comment by jaskarth** [comment]
Thanks a lot for testing, and thanks all for reviews and feedback with this change!

/integrate
